### ISSUE TITLE
captureX() methods should call their callbacks after data has been successfully logged

### ISSUE THREAD

**Issue by 256dpi** [getsentry/raven-node#86]
The methods `captureError`, `captureMessage` and `captureQuery` call their callbacks to early: https://github.com/getsentry/raven-node/blob/master/lib/client.js#L128

In my code I use error domains to capture errors:

```javascript
var d = domain.create();

domain.run(function(){
  // code that throws errors
});
```

However if I use the `captureException` method, the events do not arrive at getsentry:

```javascript
domain.on('error', function(){
  sentry.captureException(err, function(){
    process.exit(1);
  });
});
```

The `process.exit` call will terminate the process to early. I think the behaviour of these methods should be similar to `patchGlobal` which calls its callback when the client emits the `logged` event.

Am I right, or did I miss something?

**Comment by jbraithwaite** [comment]
You're right, it took me a moment to figure out why sentry wasn't working.

**Comment by shamasis** [comment]
Same here :-1: 

The following code *does not* work... for obvious reasons:

```javascript
client.captureError(err, {}, function () {
    process.exit(1);
});
```

The only solution I could do temporarily is to set a timeout hoping that log would get sent in that time:

```javascript
client.captureError(err, {}, function () {
	setTimeout(function () {
		process.exit(1);
	}, 1000);
});
```

**Comment by shamasis** [comment]
I sent PR: https://github.com/getsentry/raven-node/pull/125 to address this. Without this fix, this module is virtually useless for those who wants to terminate the process post calling `.capture*` function.

**Comment by mattrobenolt** [comment]
I documented my reasoning for this pattern over in #125, so closing this here.

**Comment by pekeler** [comment]
I just ran into this issue after this used to work for me when I tested this previously. I guess the network used to be faster for me than the filesystem. Today, I've been stumped during testing why my errors don't show up on getsentry. Now I'm reading this issue and also #125 and I'm frankly wondering how many exceptions did I not get on getsentry from my production environment due to this callback behaviour?

The main reason anybody would be using Sentry and raven is to get notified of errors and having them in one place. How could I possibly prioritize this less than getting some ID? Isn't that ID completely useless without a corresponding error in getsentry?

Using the `logged` event can get ugly quickly when I need to do different things based on the context.

As a first step, may I suggest to update the documentation. It **needs** to be made explicit that captureException is an asynchronous function and that it's callback may be called before the network transfer is complete.

After that, either

- change the api so that ID generating and error sending are two explicit functions that I need to chain
- or change captureException so it first get's the ID, then makes the transfer, then calls my callback
- or add a second callback parameter for the transfer to captureException


I have read #125 but have to fundamentally disagree with @mattrobenolt reason. Capturing the error on getsentry is so much more important

**Comment by pekeler** [comment]
As @256dpi mentioned, the inconsistency with patchGlobal's callback is also unfortunate.

OK, let's say I always want to exit, so I could use an even handler for the `logged` event to exit in that one. Now with that in place, patchGlobal's callback is never invoked and I have no clean way of logging uncaught errors.

**Comment by jishi** [comment]
I feel like this should be revisited. It feels kind of useless to have a callback api that doesn't fire when stuff is actually finished. Also, most of the stuff that happens is synchronous, it is only the error parsing that is asynchronous. 

Having to listen on events in order to know if your errors has been reported upstream is super awkward and very counter intuitive.

I think the question that should be answered is, "Who would like to know when the error has been parsed by raven and do something useful with it?" I'm guessing no one, since it was your own code that triggered it and you already have the error. How it is parsed and prepared for sending it to the server is completely internal and probably of no use.

**Comment by benvinegar** [comment]
I'm going to re-open this in order to revisit. I think there are a bunch of potential solutions here.

**Comment by LewisJEllis** [comment]
Addressed in #217.
